src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,161 @@
+#include "NoteEncryption.hpp"
+#include "crypto/sha512.h"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+#include "zerocash/Zerocash.h"
+
+#define NOTEENCRYPTION_CIPHER_KEYSIZE 32
+
+void clamp_curve25519(unsigned char *key)
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &seed,
+    unsigned char nonce
+   )
+{
+    unsigned char K_tmp[64];
+    CSHA512 hasher;
+
+    // Trim the first byte from seed
+    std::vector<unsigned char> first_248(seed.begin() + 1, seed.end());
```

Comment:
The spec is going to change to use blake2, but the spec I'm looking at while I'm reviewing this PR says trailing246(hSig), which is wrong since that's 10 bits less than 256 bits. Filed https://github.com/zcash/zips/issues/31